okio/src/main/java/okio/ByteString.java

```diff
@@ -211,6 +211,37 @@ public ByteString toAsciiUppercase() {
     return this;
   }
 
+  /**
+   * Returns a byte string that is a substring of this byte string, beginning at the specified 
+   * index until the end of this string. Returns this byte string if {@code beginIndex} is 0.
+   */
+  public ByteString substring(int beginIndex) {
+    return substring(beginIndex, data.length);
+  }
+
+  /**
+   * Returns a byte string that is a substring of this byte string, beginning at the specified 
+   * {@code beginIndex} and ends at the specified {@code endIndex}. Returns this byte string if 
+   * {@code beginIndex} is 0 and {@code endIndex} is the length of this byte string.
+   */
+  public ByteString substring(int beginIndex, int endIndex) {
+    if (beginIndex < 0) throw new IllegalArgumentException("beginIndex < 0");
+    if (endIndex > data.length) {
+      throw new IllegalArgumentException("endIndex > length(" + data.length + ")");
+    }
+    
+    int subLen = endIndex - beginIndex;
+    if (subLen < 0) throw new IllegalArgumentException("endIndex < beginIndex");
+    
+    if ((beginIndex == 0) && (endIndex == data.length)) {
+      return this;
+    }
+
+    byte[] copy = new byte[subLen];
+    System.arraycopy(data, beginIndex, copy, 0, subLen);
```

Comment:
I think there's advantages in re-using the underlying storage. For example, we read headers using a `ByteString` and it would be advantageous to do a zero-copy "split" it into name and value. Perhaps we document the fact that it's a view and require you to call `.clone()` if you want a defensive copy?